File: Problems/No.91_Decode_Ways/decode_ways.py
class Solution(object):
    def numDecodings(self, s):
        """
        :type s: str
        :rtype: int
        """

        # parameters
        s_length = len(s)
        DP = [0] * (s_length+1)

        # start computing
        DP[0] = 1
        DP[1] = 0 if (s[0] == '0') else 1
        for i in range(2, s_length+1):
            if (s[i-1] != '0'):
                DP[i] += DP[i-1]
            if ((s[i-2] == '1') or ((s[i-2] == '2') and (s[i-1] <= '6'))):
                DP[i] += DP[i-2]
        
        return DP[s_length]

File: Problems/No.91_Decode_Ways/decode_ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        
        // parameters
        int s_length = s.length();
        std::vector<int> DP(s_length+1, 0);

        // start computing
        DP[0] = 1;
        for (int i=1; i<=s_length; i++) {

            if (s[i-1] != '0') DP[i] += DP[i-1];
            if ((1 < i) && ((s[i-2] == '1') || ((s[i-2] == '2') && (s[i-1] <= '6')))) 
                DP[i] += DP[i-2];

        }
        
        return DP[s_length];

    }
};
